src/Backups/BackupCoordinationStageSync.h

```diff
@@ -10,33 +10,164 @@ class BackupCoordinationStageSync
 {
 public:
     BackupCoordinationStageSync(
-        const String & root_zookeeper_path_,
-        WithRetries & with_retries_,
+        bool is_restore_,                    /// true if this is a RESTORE ON CLUSTER command, false if this is a BACKUP ON CLUSTER command
+        const String & zookeeper_path_,      /// path to the "stage" folder in ZooKeeper
+        const String & current_host_,        /// the current host, or an empty string if it's the initiator of a BACKUP/RESTORE ON CLUSTER command
+        bool allow_concurrency_,             /// whether it's allowed to have concurrent backups or restores.
+        const WithRetries & with_retries_,
+        ThreadPoolCallbackRunnerUnsafe<void> schedule_,
```

Comment:
The main change of this PR is that this `BackupCoordinationStageSync` was completely rewritten. This class creates some nodes in ZooKeeper and also tracks such nodes created by other hosts which are also executing the same `BACKUP ON CLUSTER` or `RESTORE ON CLUSTER` command. There are multiple nodes - "started*", "current*", "alive*", and "finished*".
The "alive*" nodes are ephemeral and they're used to tell other hosts that the current host is still working on the BACKUP/RESTORE operation. Earlier those "alive" were recreated and checked from time to time quite randomly which causes weird errors. In this PR I changed it - now there is a separate thread which helps to both recreate those "alive*" nodes quickly and to react to an error happened on other host quickly. A cancellation with `KILL QUERY` is handled as error `QUERY_WAS_CANCELLED` so as an another important result we've got an ability to cancel on cluster backups and restores.